document/document.go

```diff
@@ -67,15 +67,78 @@ func (fb FieldBuffer) GetByField(field string) (Value, error) {
 	return Value{}, ErrFieldNotFound
 }
 
+// treatArrayValue update the value of array at the given index.
+func treatArrayValue(vlist Array, v Value, index int) (ValueBuffer, error) {
+	var buf ValueBuffer
+	err := vlist.Iterate(func(i int, va Value) error {
+		if index == i {
+			buf = buf.Append(v)
+			return nil
+		}
+		buf = buf.Append(va)
+		return nil
+	})
+
+	return buf, err
+}
+
+// SetDotNotation allow dot notation and replace value at the given index.
+func (fb *FieldBuffer) SetDotNotation(fname string, v Value) {
```

Comment:
I think this logic could be written in the [`Set`](https://github.com/genjidb/genji/blob/b4c31313424dfe91b04a3118225ce1a39a8322a1/document/document.go#L220) method. Basically, the `Set` method is what was used by the UPDATE logic, it is used to only support top-level field modification.
Since your PR changes that, I think modifying `Set` rather than creating an alternative method is the way to go. What do you think?

Reply:
Regarding the signature of the `Set` method, I think we should explicitly expect a `ValuePath` instead of a string. That way we don't have to cast fname to vpath, or to use the `isDotPath` function.

Reply:
I discover `ValuePath ` and it  was a reflexion about change but I  hesitate to change the existing code.I am agree with you, I think `Set` method can be change and return an error if we do the `buf.Copy(..)` inside.